framework/src/userobjects/UserObject.C

```diff
@@ -93,7 +93,20 @@ UserObject::UserObject(const InputParameters & parameters)
   mooseAssert(_sys.varKind() == Moose::VAR_NONLINEAR,
               "We expect the system to conceptually be nonlinear");
 
+  // insert self
   _supplied_uo.insert(name());
+
+  // add items for dependency resolution of writable coupled variables
+  if (auto coupleable = dynamic_cast<Coupleable *>(this); coupleable)
+  {
+    // add writable coupled variables as supplied
+    for (const auto & var : coupleable->getWritableCoupledVariables())
+      _supplied_uo.insert("Variable/" + var->name());
```

Comment:
Man, I wish this resolution was done with pointers to MooseObjects instead of strings :(

Reply:
Two issues with that:
1. thread copies of objects. But there might be a way to always get a pointer to the TID=0 object. 
2. things that are not MooseObjects like MaterialProperties

Either way, I completely agree, strings is not the way to go.